File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/scan/DocumentCaptureSurface.kt
/*
 * DocumentCaptureSurface.kt
 *
 * The Document branch of the capture screen — preserves the
 * existing ML Kit `GmsDocumentScanning` flow verbatim. Owned by
 * [QuickCaptureScreen], which mounts this composable when
 * [CaptureMode.Document] is active and tears it down on a flip
 * to [CaptureMode.BusinessCard].
 *
 * Behavior is the same as the previous all-in-one QuickCaptureScreen
 * body: page-mode pill (Single/Multi-page/Auto) + tilted lined-
 * paper page mock + shutter that launches the system scanner +
 * Import button that opens the system photo picker. The scanner
 * result flows through `ScanFlowController.onScanComplete` exactly
 * as before — no detector swap, no overlay, no in-app camera.
 *
 * Renamed from the original private `CaptureMode` (Single /
 * MultiPage / Auto) to [ScanPageMode] so the new top-level
 * [CaptureMode] can take the canonical name. The pill labels and
 * `pageLimit` plumbing are unchanged.
 *
 * Mirror of iOS `DocumentCaptureSurface.swift`.
 */

package app.quickink.mobile.features.scan

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bolt
import androidx.compose.material.icons.filled.Image
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing
import app.releaf.shared.scan.rememberDocumentScannerLauncher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Page-count intent passed to ML Kit's `setPageLimit`. Was named
 * `CaptureMode` historically; renamed to make room for the top-
 * level [CaptureMode] enum that picks between Document and
 * Business Card surfaces.
 */
internal enum class ScanPageMode(val label: String) {
    /** Single-shot capture — `pageLimit = 1` hides Add-page. */
    Single("Single"),
    /** Multi-page — no `pageLimit`, in-scanner Add-page visible. */
    MultiPage("Multi-page"),
    /** Auto — same scanner config as Multi-page today. */
    Auto("Auto"),
}

@Composable
internal fun DocumentCaptureSurface(
    controller: ScanFlowController,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current

    var pageMode by remember { mutableStateOf(ScanPageMode.Single) }
    // Page-count badge in Multi-page — reflects the LAST scanner
    // session's page count, reset on mode change.
    var pageCount by remember { mutableIntStateOf(0) }

    LaunchedEffect(pageMode) {
        pageCount = 0
    }

    val scannerLauncher = rememberDocumentScannerLauncher(
        onResult = { result ->
            pageCount = result.pageUris.size
            // Document surface fires a manual capture event whenever
            // the shutter actually returns a result. Auto-capture
            // doesn't apply here — that's a Business Card-only
            // behavior — so every Document path is "manual".
            CaptureAnalytics.manualFired(CaptureMode.Document)
            controller.onScanComplete(result)
            onDismiss()
        },
        onError = { /* TODO surface error */ },
        pageLimit = if (pageMode == ScanPageMode.Single) 1 else null,
        galleryImportAllowed = false,
    )

    val context = LocalContext.current
    val importScope = rememberCoroutineScope()
    val importLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.PickMultipleVisualMedia(),
    ) { uris ->
        if (uris.isEmpty()) return@rememberLauncherForActivityResult
        importScope.launch {
            val result = withContext(Dispatchers.IO) {
                buildImportArtifacts(context, uris)
            } ?: return@launch
            pageCount = result.pageUris.size
            controller.onScanComplete(result, source = "import")
            onDismiss()
        }
    }

    val transition = rememberInfiniteTransition(label = "scan-sweep")
    val sweep by transition.animateFloat(
        initialValue = -50f,
        targetValue  = 130f,
        animationSpec = infiniteRepeatable(
            animation = tween(1400, easing = LinearEasing),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "scan-sweep-y",
    )

    Column(
        modifier = modifier.fillMaxSize(),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Spacer(Modifier.weight(1f))

        Box(contentAlignment = Alignment.Center) {
            Box(
                modifier = Modifier
                    .rotate(-4f)
                    .size(width = 240.dp, height = 320.dp)
                    .shadow(24.dp, RoundedCornerShape(QuickInkRadius.md))
                    .clip(RoundedCornerShape(QuickInkRadius.md))
                    .background(colors.surface),
            ) {
                Canvas(modifier = Modifier.fillMaxSize().padding(vertical = 14.dp)) {
                    val lineColor = colors.ink.copy(alpha = 0.10f)
                    val spacing = 18.dp.toPx()
                    var y = spacing
                    while (y < size.height) {
                        drawLine(
                            color       = lineColor,
                            start       = Offset(0f, y),
                            end         = Offset(size.width, y),
                            strokeWidth = 0.5f,
                        )
                        y += spacing
                    }
                }
                Box(
                    modifier = Modifier
                        .padding(start = 24.dp, top = 14.dp, bottom = 14.dp)
                        .width(1.5.dp)
                        .fillMaxSize()
                        .background(colors.accent.copy(alpha = 0.6f))
                )
                Text(
                    text     = "Brainstorm — Q3\nGoals, opportunities,\nand notes…",
                    style    = type.handwritten.copy(fontSize = 20.sp),
                    color    = colors.ink.copy(alpha = 0.78f),
                    modifier = Modifier.padding(start = 36.dp, top = QuickInkSpacing.s5, end = QuickInkSpacing.s4),
                )
            }

            Box(
                modifier = Modifier
                    .offset(y = sweep.dp)
                    .size(width = 240.dp, height = 2.dp)
                    .background(
                        Brush.horizontalGradient(
                            listOf(
                                colors.accent.copy(alpha = 0f),
                                colors.accent.copy(alpha = 0.7f),
                                colors.accent,
                                colors.accent.copy(alpha = 0.7f),
                                colors.accent.copy(alpha = 0f),
                            )
                        )
                    ),
            )

            DocumentCornerMark(rotation = 0f,   xOffset = (-120).dp, yOffset = (-160).dp)
            DocumentCornerMark(rotation = 90f,  xOffset =  120.dp,   yOffset = (-160).dp)
            DocumentCornerMark(rotation = 270f, xOffset = (-120).dp, yOffset =  160.dp)
            DocumentCornerMark(rotation = 180f, xOffset =  120.dp,   yOffset =  160.dp)
        }

        Spacer(Modifier.weight(1f))

        // Shutter row — page badge / shutter / import.
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = QuickInkSpacing.s5, vertical = QuickInkSpacing.s4),
            verticalAlignment     = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Box(modifier = Modifier.size(width = 64.dp, height = 64.dp), contentAlignment = Alignment.Center) {
                if (pageMode == ScanPageMode.MultiPage) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text(
                            text  = "$pageCount",
                            style = type.heading,
                            color = Color.White,
                        )
                        Text(
                            text  = if (pageCount == 1) "page" else "pages",
                            style = type.caption,
                            color = Color.White.copy(alpha = 0.7f),
                        )
                    }
                }
            }

            DocumentShutterButton(onClick = scannerLauncher::launch)

            Box(
                modifier = Modifier.size(width = 64.dp, height = 64.dp),
                contentAlignment = Alignment.Center,
            ) {
                DocumentImportButton(
                    onClick = {
                        importLauncher.launch(
                            PickVisualMediaRequest(
                                ActivityResultContracts.PickVisualMedia.ImageOnly
                            )
                        )
                    },
                )
            }
        }

        // Page-mode pill (Single / Multi-page / Auto). Lives below
        // the shutter row, unchanged from the pre-refactor layout.
        Row(
            modifier = Modifier
                .padding(horizontal = QuickInkSpacing.s5)
                .padding(bottom = QuickInkSpacing.s7)
                .fillMaxWidth(),
            horizontalArrangement = Arrangement.Center,
        ) {
            Row(
                modifier = Modifier
                    .clip(RoundedCornerShape(QuickInkRadius.pill))
                    .background(Color.White.copy(alpha = 0.10f))
                    .padding(4.dp),
            ) {
                ScanPageMode.values().forEachIndexed { i, m ->
                    if (i > 0) Spacer(Modifier.size(4.dp))
                    val active = (m == pageMode)
                    Box(
                        modifier = Modifier
                            .clip(RoundedCornerShape(QuickInkRadius.pill))
                            .background(if (active) colors.accent else Color.Transparent)
                            .clickable { pageMode = m }
                            .padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2),
                    ) {
                        Text(
                            text  = m.label,
                            style = type.label,
                            color = if (active) Color.White else Color.White.copy(alpha = 0.55f),
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun DocumentShutterButton(onClick: () -> Unit) {
    val colors = LocalQuickInkColors.current
    Box(
        modifier         = Modifier
            .size(78.dp)
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .size(78.dp)
                .border(3.dp, Color.White.copy(alpha = 0.6f), CircleShape),
        )
        Box(
            modifier = Modifier
                .size(64.dp)
                .shadow(16.dp, CircleShape)
                .clip(CircleShape)
                .background(colors.accent),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector       = Icons.Filled.Bolt,
                contentDescription = "Capture",
                tint              = Color.White,
                modifier          = Modifier.size(26.dp),
            )
        }
    }
}

@Composable
private fun DocumentImportButton(onClick: () -> Unit) {
    Box(
        modifier = Modifier
            .size(48.dp)
            .clip(CircleShape)
            .background(Color.White.copy(alpha = 0.10f))
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector       = Icons.Filled.Image,
            contentDescription = "Import photo",
            tint              = Color.White.copy(alpha = 0.85f),
            modifier          = Modifier.size(22.dp),
        )
    }
}

@Composable
private fun DocumentCornerMark(
    rotation: Float,
    xOffset: Dp,
    yOffset: Dp,
) {
    val colors = LocalQuickInkColors.current
    Canvas(
        modifier = Modifier
            .offset(x = xOffset, y = yOffset)
            .rotate(rotation)
            .size(18.dp),
    ) {
        val path = Path().apply {
            moveTo(0f, size.height)
            lineTo(0f, 0f)
            lineTo(size.width, 0f)
        }
        drawPath(
            path  = path,
            color = colors.accent,
            style = Stroke(width = 3f.dp.toPx(), cap = StrokeCap.Round),
        )
    }
}

File: apps/quickink/ios/QuickInk/Scan/DocumentCaptureSurface.swift
/*
 * DocumentCaptureSurface.swift
 *
 * The Document branch of the capture screen — preserves the
 * existing VisionKit `VNDocumentCameraViewController` flow
 * verbatim. Owned by `QuickCaptureScreen`, which mounts this
 * view when `CaptureMode.document` is active and tears it
 * down on a flip to `.businessCard` (pill), `.photo`
 * (shutter-row camera icon tap), or `.video` (shutter-row
 * camera icon long-press).
 *
 * Behavior is the same as the previous all-in-one
 * QuickCaptureScreen body: page-mode pill (Single /
 * Multi-page) + tilted lined-paper page mock + shutter
 * that presents the system scanner + Import button that opens
 * PhotosPicker. The shutter row's left slot now hosts a
 * camera icon: tap fires `onSelectPhoto`, long-press fires
 * `onSelectVideo`. The scanner result flows through
 * `ScanFlowController.onScanComplete` exactly as before — no
 * detector swap, no overlay, no in-app camera.
 *
 * Renamed from the original nested `enum CaptureMode`
 * (Single / Multi-page) to [ScanPageMode] so the new
 * top-level `CaptureMode` can take the canonical name. The
 * pill labels and `pageLimit` plumbing are unchanged.
 *
 * Mirror of Android `DocumentCaptureSurface.kt`.
 */

import SwiftUI
import PhotosUI
import ReleafCoreScan

enum ScanPageMode: String, CaseIterable {
    case single, multiPage

    var label: String {
        switch self {
        case .single:    return "Single"
        case .multiPage: return "Multi-page"
        }
    }
}

struct DocumentCaptureSurface: View {
    let controller: ScanFlowController
    let onDismiss: () -> Void
    /// Callback fired when the user taps the Photo icon in the
    /// shutter row's left slot. Threaded up through the
    /// `QuickCaptureScreen` parent so the coordinator (which
    /// owns mode state) can flip to `.photo` and SwiftUI swaps
    /// in `PhotoCaptureSurface`. Defaults to a no-op so callers
    /// that don't host a coordinator (previews) keep compiling.
    var onSelectPhoto: () -> Void = {}
    /// Long-press counterpart to `onSelectPhoto`; starts the
    /// dedicated Video surface from the same camera icon.
    var onSelectVideo: () -> Void = {}

    @State private var pageMode: ScanPageMode = .single
    @State private var sweepOffset: CGFloat = -50
    @State private var showSystemScanner = false
    @State private var pickedItems: [PhotosPickerItem] = []

    var body: some View {
        VStack(spacing: 0) {
            Spacer()
            pagePreview
            Spacer()
            shutterRow
            pageModeSelector
                .padding(.bottom, QuickInkSpacing.s7)
        }
        .onChange(of: pickedItems) { newItems in
            guard !newItems.isEmpty else { return }
            Task { await loadPickedItems(newItems) }
        }
        .fullScreenCover(isPresented: $showSystemScanner) {
            DocumentScannerView(
                onComplete: { pdfURL, previewURL, pageURLs in
                    showSystemScanner = false
                    CaptureAnalytics.manualFired(mode: .document)
                    controller.onScanComplete(
                        pdfURL:     pdfURL,
                        previewURL: previewURL,
                        pageURLs:   pageURLs,
                    )
                    onDismiss()
                },
                onCancel: { showSystemScanner = false },
                pageLimit: pageMode == .single ? 1 : nil,
                compressedPdfEnabled: SettingsState.compressedPdfSavesDefault,
            )
            .ignoresSafeArea()
        }
    }

    // MARK: - Page preview

    @ViewBuilder
    private var pagePreview: some View {
        ZStack {
            ZStack(alignment: .topLeading) {
                QuickInkLinedPaper(
                    tone: QuickInkColors.surface,
                    lineSpacing: 18,
                    lineOpacity: 0.10,
                )
                HStack(spacing: 0) {
                    Rectangle()
                        .fill(QuickInkColors.accent.opacity(0.6))
                        .frame(width: 1.5)
                        .padding(.leading, 24)
                    Spacer()
                }
                .padding(.vertical, 14)

                Text("Brainstorm — Q3\nGoals, opportunities,\nand notes…")
                    .font(QuickInkFont.handwritten(20))
                    .foregroundStyle(QuickInkColors.ink.opacity(0.78))
                    .padding(.leading, 36)
                    .padding(.top, QuickInkSpacing.s5)
            }
            .frame(width: 240, height: 320)
            .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
            .rotation3DEffect(.degrees(8), axis: (x: 1, y: 0, z: 0))
            .rotation3DEffect(.degrees(-4), axis: (x: 0, y: 1, z: 0))
            .shadow(color: .black.opacity(0.5), radius: 24, x: 0, y: 12)

            Rectangle()
                .fill(
                    LinearGradient(
                        colors: [
                            QuickInkColors.accent.opacity(0),
                            QuickInkColors.accent.opacity(0.7),
                            QuickInkColors.accent,
                            QuickInkColors.accent.opacity(0.7),
                            QuickInkColors.accent.opacity(0),
                        ],
                        startPoint: .leading, endPoint: .trailing,
                    )
                )
                .frame(width: 240, height: 2)
                .offset(y: sweepOffset)
                .shadow(color: QuickInkColors.accent.opacity(0.6), radius: 8, x: 0, y: 0)

            DocumentDetectionCorner(rotation: 0)        .offset(x: -120, y: -160)
            DocumentDetectionCorner(rotation: 90)       .offset(x:  120, y: -160)
            DocumentDetectionCorner(rotation: 270)      .offset(x: -120, y:  160)
            DocumentDetectionCorner(rotation: 180)      .offset(x:  120, y:  160)
        }
        .onAppear {
            withAnimation(.easeInOut(duration: 1.4).repeatForever(autoreverses: true)) {
                sweepOffset = 130
            }
        }
    }

    // MARK: - Shutter row

    @ViewBuilder
    private var shutterRow: some View {
        HStack(alignment: .center) {
            // Left slot — Photo icon (was a spacer pre-Photo
            // capture). Same 48pt disc styling as the right-
            // side import button so the row reads as a
            // symmetric two-button frame around the shutter.
            photoModeButton.frame(width: 64, height: 64)
            Spacer()
            documentShutterButton
            Spacer()
            importButton.frame(width: 64, height: 64)
        }
        .padding(.horizontal, QuickInkSpacing.s5)
        .padding(.bottom, QuickInkSpacing.s4)
    }

    @ViewBuilder
    private var photoModeButton: some View {
        Image(systemName: "camera")
            .font(.system(size: 22, weight: .medium))
            .foregroundStyle(Color.white.opacity(0.85))
            .frame(width: 48, height: 48)
            .background(Color.white.opacity(0.10))
            .clipShape(Circle())
            .contentShape(Circle())
            .gesture(
                LongPressGesture(minimumDuration: 0.45)
                    .exclusively(before: TapGesture())
                    .onEnded { value in
                        switch value {
                        case .first(true):
                            UIImpactFeedbackGenerator(style: .medium).impactOccurred()
                            onSelectVideo()
                        case .second:
                            UIImpactFeedbackGenerator(style: .light).impactOccurred()
                            onSelectPhoto()
                        default:
                            break
                        }
                    }
            )
        .accessibilityLabel("Take a photo")
        .accessibilityHint("Long press to record a video")
    }

    @ViewBuilder
    private var documentShutterButton: some View {
        Button(action: { showSystemScanner = true }) {
            ZStack {
                Circle()
                    .stroke(Color.white.opacity(0.6), lineWidth: 3)
                    .frame(width: 78, height: 78)
                Circle()
                    .fill(QuickInkColors.accent)
                    .frame(width: 64, height: 64)
                Image(systemName: "bolt.fill")
                    .font(.system(size: 26, weight: .semibold))
                    .foregroundStyle(.white)
            }
            .shadow(color: QuickInkColors.accent.opacity(0.5), radius: 16, x: 0, y: 6)
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Capture page")
    }

    @ViewBuilder
    private var importButton: some View {
        PhotosPicker(
            selection:    $pickedItems,
            matching:     .images,
            photoLibrary: .shared(),
        ) {
            Image(systemName: "photo.on.rectangle")
                .font(.system(size: 22, weight: .medium))
                .foregroundStyle(Color.white.opacity(0.85))
                .frame(width: 48, height: 48)
                .background(Color.white.opacity(0.10))
                .clipShape(Circle())
        }
        .accessibilityLabel("Import photos")
    }

    // MARK: - Page-mode selector

    @ViewBuilder
    private var pageModeSelector: some View {
        HStack(spacing: 4) {
            ForEach(ScanPageMode.allCases, id: \.self) { m in
                let active = (m == pageMode)
                Button(action: { pageMode = m }) {
                    Text(m.label)
                        .font(QuickInkText.label)
                        .foregroundStyle(active ? .white : Color.white.opacity(0.55))
                        .padding(.horizontal, QuickInkSpacing.s4)
                        .padding(.vertical, QuickInkSpacing.s2)
                        .background(
                            RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous)
                                .fill(active ? QuickInkColors.accent : .clear)
                        )
                }
                .buttonStyle(.plain)
            }
        }
        .padding(4)
        .background(Color.white.opacity(0.10))
        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
    }

    // MARK: - Photo import

    private func loadPickedItems(_ items: [PhotosPickerItem]) async {
        var images: [UIImage] = []
        images.reserveCapacity(items.count)
        for item in items {
            guard let data  = try? await item.loadTransferable(type: Data.self),
                  let image = UIImage(data: data) else {
                continue
            }
            images.append(image)
        }
        guard let result = ImportArtifacts.build(from: images) else {
            await MainActor.run { pickedItems = [] }
            return
        }
        await MainActor.run {
            controller.onScanComplete(
                pdfURL:     result.pdfURL,
                previewURL: result.previewURL,
                pageURLs:   result.pageURLs,
                source:     "import",
            )
            pickedItems = []
            onDismiss()
        }
    }
}

private struct DocumentDetectionCorner: View {
    let rotation: Double
    var body: some View {
        Path { path in
            path.move(to: CGPoint(x: 0, y: 18))
            path.addLine(to: CGPoint(x: 0, y: 0))
            path.addLine(to: CGPoint(x: 18, y: 0))
        }
        .stroke(QuickInkColors.accent, style: StrokeStyle(lineWidth: 3, lineCap: .round, lineJoin: .round))
        .frame(width: 18, height: 18)
        .rotationEffect(.degrees(rotation))
    }
}
